### PR TITLE
Refactor Catalog Indexing

### PR DESCRIPTION
**Pull request by ramonski**
## Description of the issue/feature this PR addresses

This PR refactors the catalog indexing by using the `IndexQueue` for all SENAITE catalogs and avoids multiple UID catalog indexing.

## Current behavior before PR

AT object indexing never use the IndexQueue and temporary objects were indexed in UID catalog

## Desired behavior after PR is merged

AT object indexing always use the IndexQueue and no temporary objects are indexed in UID catalog

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/bika/lims/utils/analysisrequest.py [hunk 1]
```diff
@@ -175,6 +175,10 @@ def reindex(obj, recursive=False):
     :param obj: The object to reindex
     :param recursive: If true, all child objects are reindexed recursively
     """
+    # reindex UID
+    uid_catalog = api.get_tool("uid_catalog")
+    uid_catalog.catalog_object(obj, obj._getURL())
+    # reindex object in all other catalogs
```
Comment (xispa) [inline src/bika/lims/utils/analysisrequest.py]: To not spread here and there same thing, I suggest to rely on `api.catalog_object` function, that already takes care of `uid_catalog` here: https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/api/__init__.py#L403-L417

We could even include the `recursive` param in `api.catalog_object` function as well and remove this `reindex` function from here. What do you think?

File: src/senaite/core/catalog/catalog_multiplex_processor.py [hunk 2]
```diff
@@ -46,7 +46,11 @@ def is_global_auditlog_enabled(self):
         return setup.getEnableGlobalAuditlog()
 
     def get_catalogs_for(self, obj):
-        catalogs = getattr(obj, "_catalogs", [])
+        """Get a list of catalog IDs for the given object
+        """
+        # get a list of catalog IDs that are mapped to the object
+        catalogs = list(map(lambda x: x.id, api.get_catalogs_for(obj)))
```
Comment (xispa) [inline src/senaite/core/catalog/catalog_multiplex_processor.py:52]: Note that by default `api.get_catalogs_for` relies on [this static mapping](https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/catalog/__init__.py#L49) to get the catalogs for a given portal type. And the value of the attribute `_catalog` is taken into account only when the portal type of the given object is not present in that mapping: https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/api/__init__.py#L1276

In my opinion this *static mapping* forces us to monkey patch `api.get_catalogs_for` if we want an existing baseline type (like AnalysisRequest) to also be catalogued in a brand-new-add-on-specific catalog.

So, I think this change is fine as long as we skip the *static mapping* check in `api.get_catalogs_for`. Or maybe don't skip, but rely on a new adapter to get  additional catalogs for baseline portal_types.

File: src/senaite/core/patches/archetypes/referenceable.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+from Products.Archetypes.config import UID_CATALOG
+
+
+def _catalogUID(self, aq, uc=None):
+    # skip indexing of temporary objects
+    if api.is_temporary(self):
+        return
+    if not uc:
+        uc = api.get_tool(UID_CATALOG)
+    url = self._getURL()
```
Comment (xispa) [inline src/senaite/core/patches/archetypes/referenceable.py:31]: Note for @xispa : `._getURL()` relies on `Archetypes.utils.getRelPath`, that returns the relative path of the object from the portal root. The result is the same as `url = "/".join(obj.getPhysicalPath()[2:])`

File: src/senaite/core/patches/cmfcore/portal_catalog_processor.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#
+# Copyright 2018-2024 by it's authors.
+# Some rights reserved, see README and LICENSE.
+
+from bika.lims import api
+
+PORTAL_CATALOG = "portal_catalog"
+
+
+def index_in_portal_catalog(obj):
```
Comment (xispa) [inline src/senaite/core/patches/cmfcore/portal_catalog_processor.py:26]: Shouldn't we also check for `api.is_temporary(obj)`, like what you did in `catalog_multiplex`?

**Review by xispa — CHANGES_REQUESTED** [review]
Awesome!, thanks for the doctest and README's (really useful and necessary). Have left some comments and questions

File: src/bika/lims/utils/analysisrequest.py (hunk 1 above)
Comment (ramonski) [reply src/bika/lims/utils/analysisrequest.py]: Thanks for the hint! I already forgot that we have an API function that does this.
Changes done in [c725d9d](https://github.com/senaite/senaite.core/pull/2632/commits/c725d9d0c0652fd0e0b013f16154ede0be6d5b99)

File: src/senaite/core/catalog/catalog_multiplex_processor.py (hunk 2 above)
Comment (ramonski) [reply src/senaite/core/catalog/catalog_multiplex_processor.py:52]: Yes, you are right and this is why we have the `TODO` in the docstring here:
https://github.com/senaite/senaite.core/blob/c725d9d0c0652fd0e0b013f16154ede0be6d5b99/src/senaite/core/catalog/__init__.py#L107-L120

I guess providing the mapping as a registry setting we could easily add/remove/edit catalogs from the base types and new types.

For the moment I would like to postpone that until we have this requirement and provide it in a new PR

File: src/senaite/core/patches/cmfcore/portal_catalog_processor.py (hunk 4 above)
Comment (ramonski) [reply src/senaite/core/patches/cmfcore/portal_catalog_processor.py:26]: Everything that ends up in this queue comes from `Products.CMFCore.CatalogTool.reindexObject` which does already `filterTemporaryItems`, which uses our patched `isTemporary` for both AT and DX types:

https://github.com/senaite/senaite.core/blob/c725d9d0c0652fd0e0b013f16154ede0be6d5b99/src/senaite/core/patches/archetypes/base_object.py#L29-L30

https://github.com/senaite/senaite.core/blob/c725d9d0c0652fd0e0b013f16154ede0be6d5b99/src/senaite/core/patches/dexterity/dexterity_content.py#L29-L30

So another `api.is_temporary` is at this stage not required

**Review by xispa — APPROVED** [review]
Excellent, thanks!

**Pull request merged**